contracts/src/L2/L2EjectionController.sol

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.13;
+
+import {IL2EjectionController} from "./IL2EjectionController.sol";
+import {IStandardRegistry} from "../common/IStandardRegistry.sol";
+import {ITokenObserver} from "../common/ITokenObserver.sol";
+import {IERC1155Receiver} from "@openzeppelin/contracts/token/ERC1155/IERC1155Receiver.sol";
+import {IRegistry} from "../common/IRegistry.sol";
+
+/**
+ * @title L2EjectionController
+ * @dev L2 contract for ejection controller that facilitates migrations of names
+ * between L1 and L2, as well as handling renewals.
+ */
+contract L2EjectionController is IL2EjectionController {
+    error NotTokenOwner(uint256 tokenId);
+    event NameEjectedToL1(uint256 indexed tokenId, address l1Owner, address l1Subregistry, uint64 expiry);
+    event NameMigratedToL2(uint256 indexed tokenId, address l2Owner, address l2Subregistry);
+    event NameRenewed(uint256 indexed tokenId, uint64 expires, address renewedBy);
+
+    IStandardRegistry public immutable registry;
+
+    constructor(IStandardRegistry _registry) {
+        registry = _registry;
+    }
+
+    /**
+     * @dev Called by the L2 eth registry when a user ejects a name to L1.
+     *
+     * @param tokenId The token ID of the name being ejected
+     * @param l1Owner The address that will own the name on L1
+     * @param l1Subregistry The subregistry address to use on L1
+     */
+    function ejectToL1(uint256 tokenId, address l1Owner, address l1Subregistry) external {
+        if (registry.ownerOf(tokenId) != address(this)) {
+            revert NotTokenOwner(tokenId);
+        }
+
+        uint64 expiry = registry.getExpiry(tokenId);
+
+        registry.setSubregistry(tokenId, IRegistry(address(0)));
+
+        // bridge will listen for this event        
```

Comment:
Shouldn't this be calling Linea's `L2MessageService.sendMessage` according to their doc https://docs.linea.build/technology/message-service ?

Reply:
As discussed, we'll probably have a pluggable bridge system, leaving the ejection controller itself to not have to worry about this.